### ISSUE TITLE
EC_POINT_cmp returns error when using a dynamic ENGINE and statically linked openssl

### ISSUE THREAD

**Issue by JackOfMostTrades** [openssl/openssl#5475]
If I create a custom engine that implements ENGINE_load_private_key and returns an EC key and use this engine dynamically, its public portion cannot be compared to a public key loaded by OpenSSL when the application is compiled against OpenSSL statically. This is because the comparison [here](https://github.com/openssl/openssl/blob/3c7d0945b655a71a026bd0d855943037d8ce2bec/crypto/ec/ec_lib.c#L866) will always be false (and thus the EC_R_INCOMPATIBLE_OBJECTS will be set).

This came up because I'm trying to use a dynamically-loaded engine in an application that has statically linked OpenSSL. The following is the skeleton for a reproducer:

```
ENGINE *engine = ENGINE_by_id("myengine.so");
EVP_PKEY *pkey = ENGINE_load_private_key(engine, "mykey", NULL, NULL); // Loads an regular, old-fashioned EC private key corresponding to the public key in the "mycert.crt" file.
SSL_CTX *ctx = SSL_CTX_new(TLS_method());
SSL_CTX_use_certificate_chain_file(ctx, "mycert.crt"); // X509 cert using the EC pub key.
if (SSL_CTX_use_PrivateKey(ctx, pkey) != 1) { // Always returns an error because the X509_check_private_key check will fail
    printf("Unable to use private key.\n");
}
```

If you'd like a full end-to-end reproducer I can try and hack something minimal together.

**Comment by davidben** [comment]
Is your engine dynamically linked with another copy of OpenSSL or statically linked with OpenSSL itself? In both those cases, mixing and matching things isn't going to work well.

Dynamic linking and static linking don't really mix in general. You have two (or more) instances of OpenSSL in your process, which means two sets of globals, two sets of locks, etc. Objects from those two OpenSSLs can't be safely passed into each other. Weird things will happen. (Consider ex_data registrations.)

You probably want to be using a dynamically-linked OpenSSL, specifically the same one the engine will ultimately load.

**Comment by JackOfMostTrades** [comment]
I run into the issue whether or not the engine is statically linked. It also happens if the app itself is dynamically linked and the engine is statically linked, or if both are statically linked (even against the exact same libssl.a/libcrypto.a). As you say, badness isn't two surprising since there's two copies of everything floating around.

That said, the whole reason this is coming up is because I'm trying to use the engine in contexts where OpenSSL is already being wrapped statically (specifically in Ruby, Python, and Node.js apps). It all works well enough (despite the multiple static linkings) when using RSA keys, but I run into this issue with EC keys.

**Comment by JackOfMostTrades** [comment]
I've made a simplified reproducer. As you can see from the below, the issue only comes up with the statically linked version of the app and it only happens for EC keys (not RSA). Also of note I came up with a hacky workaround for the issue that serializes and deserializes the EC_GROUP and public EC_POINT of the private key.

`engine.c`
```
#include <openssl/engine.h>
#include <openssl/pem.h>

static const char *engine_id = "example";
static const char *engine_name = "Example";

static int engine_init(ENGINE *e) {
    return 1;
}

static EVP_PKEY *engine_load_privkey(ENGINE *engine, const char *s_key_id,
        UI_METHOD *ui_method, void *callback_data) {
    EVP_PKEY *key;
    FILE *f = fopen(s_key_id, "r");
    if (f == NULL) {
         return NULL;
    }
    key = PEM_read_PrivateKey(f, NULL, NULL, NULL);
    fclose(f);
    return key;
}

int bind_helper(ENGINE *e, const char *id) {
    if (!ENGINE_set_id(e, engine_id) ||
            !ENGINE_set_name(e, engine_name) ||
            !ENGINE_set_init_function(e, engine_init) ||
            !ENGINE_set_load_privkey_function(e, engine_load_privkey)) {
        return 0;
    }
    return 1;
}

IMPLEMENT_DYNAMIC_CHECK_FN();
IMPLEMENT_DYNAMIC_BIND_FN(bind_helper);
```

`app.c`
```
#include <openssl/engine.h>
#include <openssl/ssl.h>

ENGINE* load_dynamic_engine(const char* so_path) {
    int rc;

    ENGINE_load_dynamic();
    ENGINE *engine = ENGINE_by_id("dynamic");
    if (engine == NULL) {
        printf("Could not load dynamic engine.\n");
        return NULL;
    }
    rc = ENGINE_ctrl_cmd_string(engine, "SO_PATH", so_path, 0);
    if (rc != 1) {
        printf("Could not load SO_PATH!\n");
        return NULL;
    }
    rc = ENGINE_ctrl_cmd_string(engine, "ID", "example", 0);
    if (rc != 1) {
        printf("Could not set id.\n");
        return NULL;
    }
    rc = ENGINE_ctrl_cmd_string(engine, "LOAD", NULL, 0);
    if (rc != 1) {
        printf("Could not send load command.\n");
        return NULL;
    }
    rc = ENGINE_init(engine);
    if (rc != 1) {
        printf("Could not init engine.\n");
        return NULL;
    }
    return engine;
}

void fix_key(EVP_PKEY *key) {
    // Proof of concept work-around for https://github.com/openssl/openssl/issues/5475
    EC_KEY *eckey;
    const EC_GROUP *group;
    int group_name;
    const EC_POINT *pubpoint;
    BN_CTX *bctx;
    size_t point_len;
    unsigned char *pointoct;
    EC_GROUP *newgroup;
    EC_POINT *newpoint;

    // Only applies to EC keys
    if (EVP_PKEY_id(key) != EVP_PKEY_EC) {
        return;
    }
    eckey = EVP_PKEY_get1_EC_KEY(key);
    group = EC_KEY_get0_group(eckey);
    group_name = EC_GROUP_get_curve_name(group);
    if (group_name == 0) {
        // If the key isn't using a named curve, this work-around won't work as-written (though we could probably reconstruct the whole group).
        return;
    }
    pubpoint = EC_KEY_get0_public_key(eckey);

    bctx = BN_CTX_new();
    BN_CTX_start(bctx);
    point_len = EC_POINT_point2oct(group, pubpoint, POINT_CONVERSION_UNCOMPRESSED, NULL, 0, bctx);
    pointoct = OPENSSL_malloc(point_len);
    EC_POINT_point2oct(group, pubpoint, POINT_CONVERSION_UNCOMPRESSED, pointoct, point_len, bctx);

    newgroup = EC_GROUP_new_by_curve_name(group_name);
    EC_KEY_set_group(eckey, newgroup);
    newpoint = EC_POINT_new(newgroup);
    EC_POINT_oct2point(newgroup, newpoint, pointoct, point_len, bctx);
    EC_KEY_set_public_key(eckey, newpoint);

    OPENSSL_free(pointoct);
    BN_CTX_end(bctx);
    BN_CTX_free(bctx);
    EC_KEY_free(eckey);
    EC_GROUP_free(newgroup);
    EC_POINT_free(newpoint);
}

void test_setup_ssl_ctx(const char *cert_path, const char *key_path, int fixkey) {
    SSL_CTX *ctx;
    EVP_PKEY *key;
    ENGINE *engine;

    engine = load_dynamic_engine("./libengine.so");
    if (engine == NULL) {
        printf("Unable to load engine.\n");
        return;
    }
    key = ENGINE_load_private_key(engine, key_path, NULL, NULL);
    if (key == NULL) {
        printf("Unable to load key.\n");
        return;
    }
    if (fixkey) {
        fix_key(key);
    }

    ctx = SSL_CTX_new(TLS_method());
    if (SSL_CTX_use_certificate_chain_file(ctx, cert_path) != 1) {
        printf("Unable to use certificate chain file.\n");
        return;
    }

    if (SSL_CTX_use_PrivateKey(ctx, key) != 1) {
        printf("Unable to use private key with (%s, %s, %d).\n", cert_path, key_path, fixkey);
    } else {
        printf("Successfully setup SSL context to use private key with (%s, %s, %d).\n", cert_path, key_path, fixkey);
    }

    EVP_PKEY_free(key);
    SSL_CTX_free(ctx);
}

int main(int argc, char** argv) {
    test_setup_ssl_ctx("rsa.crt", "rsa.key", 0);
    test_setup_ssl_ctx("ec.crt", "ec.key", 0);
    test_setup_ssl_ctx("rsa.crt", "rsa.key", 1);
    test_setup_ssl_ctx("ec.crt", "ec.key", 1);
    return 0;
}
```

`Makefile`
```
all: libengine.so app app-static

clean:
	rm -f libengine.so app app-static rsa.crt rsa.key ec.crt ec.key

libengine.so: engine.c
	gcc -shared -fPIC -Wall engine.c -o libengine.so -lssl -lcrypto

app: app.c rsa.crt ec.crt
	gcc -Wall app.c -o app -lssl -lcrypto

app-static: app.c rsa.crt ec.crt
	gcc -Wall app.c -o app-static /usr/lib64/libssl.a /usr/lib64/libcrypto.a -ldl -lpthread

rsa.crt:
	openssl req -x509 -newkey rsa:2048 -keyout rsa.key -out rsa.crt -nodes -subj '/CN=foo/'

ec.crt:
	openssl ecparam -out ec.key -name prime256v1 -genkey 
	openssl req -x509 -key ec.key -out ec.crt -nodes -subj '/CN=foo/'
```

```
$ make clean all
rm -f libengine.so app app-static rsa.crt rsa.key ec.crt ec.key
gcc -shared -fPIC -Wall engine.c -o libengine.so -lssl -lcrypto
openssl req -x509 -newkey rsa:2048 -keyout rsa.key -out rsa.crt -nodes -subj '/CN=foo/'
Generating a 2048 bit RSA private key
...............................................................................+++
................................+++
writing new private key to 'rsa.key'
-----
openssl ecparam -out ec.key -name prime256v1 -genkey 
openssl req -x509 -key ec.key -out ec.crt -nodes -subj '/CN=foo/'
gcc -Wall app.c -o app -lssl -lcrypto
gcc -Wall app.c -o app-static /usr/lib64/libssl.a /usr/lib64/libcrypto.a -ldl -lpthread
$ ./app
Successfully setup SSL context to use private key with (rsa.crt, rsa.key, 0).
Successfully setup SSL context to use private key with (ec.crt, ec.key, 0).
Successfully setup SSL context to use private key with (rsa.crt, rsa.key, 1).
Successfully setup SSL context to use private key with (ec.crt, ec.key, 1).
$ ./app-static 
Successfully setup SSL context to use private key with (rsa.crt, rsa.key, 0).
Unable to use private key with (ec.crt, ec.key, 0).
Successfully setup SSL context to use private key with (rsa.crt, rsa.key, 1).
Successfully setup SSL context to use private key with (ec.crt, ec.key, 1).
```

**Comment by davidben** [comment]
Unfortunately, I think the only plausible answer here is that statically-linked OpenSSLs and ENGINEs are not compatible. If the OpenSSL <-> ENGINE boundary were much tighter, where the ENGINE merely implements particular functions and never calls back into OpenSSL (unless through a vtable that OpenSSL manually provides), this could work. But instead OpenSSL relies on the dynamic linker to provide that vtable, which means you need to work with the dynamic linker.

Even were the particular check in question removed, you would run into problems as soon as your ENGINE used ex_data (the ENGINE presumably exists to wrap some foreign crypto API, so it needs to hang state off its EVP_PKEYs somehow).

**Comment by JackOfMostTrades** [comment]
I'm not sure I totally agree with you that you'd run into problems regardless; the whole construct of an engine is that the private key is loaded by the engine (and some extra data is hung off of ex_data), and the signing operation is also implemented by the engine (set by calling `EC_KEY_METHOD_set_sign` before returning the loaded key). If I had called `ECDSA_sign()` instead of putting the loaded private key into an SSL_CTX everything would have been fine because it would only be the the dynamic engine code doing any manipulation or introspection of the key. However, loading the private key into an SSL_CTX has the side-effect of introspecting the public portion of the key, which in this context I really don't want it to do because the key should be treated as a more opaque blob that's only meaningful to the engine.

Perhaps a better way to frame this is to request a feature enhancement that adds the EC_KEY equivalent of the `RSA_METHOD_FLAG_NO_CHECK` flag. As spelled out in an internal comment the effect of this flag is "Don't check the public/private key, this is mostly for smart cards." This is exactly the use-case I'm trying to satisfy here; inside `ssl_set_pkey` I don't want OpenSSL to check that the public component of the EVP_PKEY matches the already-loaded certificate because the key is actually an opaque blob representing a handle on a hardware token and it doesn't make sense for the OpenSSL code to try and inspect it.

**Comment by davidben** [comment]
> and some extra data is hung off of ex_data

I believe this is in itself a problem. ex_data involves a global table mapping from indices to function pointers, notably a free function. With two instances of OpenSSL in your process, you have two sets of registrations between the two copies. If one OpenSSL sees another OpenSSL's `EVP_PKEY` and attempts to set or query ex_data on it, things will go wrong. Moreover, `EVP_PKEY` is ref-counted. If the wrong OpenSSL happened to release the last reference, it will call the wrong free functions. Any of these going wrong will lead to memory errors.

To make matters worse, the `ENGINE` library itself has a global lock. Two instances of OpenSSL mean two of those locks, so I suspect some of the locking there also isn't working right.

> the EC_KEY equivalent of the RSA_METHOD_FLAG_NO_CHECK flag

There was some chatter around this floating around in #1532 and friends, but I think this particular use has more serious problems. Such a flag would, at best, resolve things on accident and depend on implementation details.

**Comment by JackOfMostTrades** [comment]
> If the wrong OpenSSL happened to release the last reference, it will call the wrong free functions.

Ok, that's definitely a fair point. To see if I correctly grok the interplay, if the ex_index was always allocated with NULL init/free functions and the ex_data cleanup was instead done in finish method of the EC_KEY (e.g. using EC_KEY_METHOD_set_init) then the cleanup would always get back to the engine no matter which OpenSSL released the last reference, right? E.g. what libp11 does for RSA keys ([here](https://github.com/OpenSC/libp11/blob/master/src/p11_rsa.c#L388) and [here](https://github.com/OpenSC/libp11/blob/master/src/p11_rsa.c#L478))?

But since that assumption shouldn't be made, I agree that there's going to be badness all over the place. Looking at #1532 and #1551, I think those might be what I really want to watch.

As a point of reference, I first started hitting this issue when using the new [clientCertEngine](https://github.com/nodejs/node/pull/14903) parameter for node.js. This allows you to point to an engine implementing `ENGINE_load_ssl_client_cert` to supply the cert/key provided when making a TLS client connection. Since node ships [binaries](https://nodejs.org/en/download/) with openssl statically linked in I started hitting this whether my engine was linked statically or dynamically. Does this suggest that this new node.js feature could lead to memory corruption even for "well-behaved" engines?

**Comment by davidben** [comment]
> if the ex_index was always allocated with NULL init/free functions and the ex_data cleanup was instead done in finish method of the EC_KEY (e.g. using EC_KEY_METHOD_set_init) then the cleanup would always get back to the engine no matter which OpenSSL released the last reference, right?

I think so, but you'd also need to worry about whether the corresponding index in the *other* OpenSSL happened to be initialized with a free function. It's also very much a "working on accident" sort of thing, based on which paths happen to or not to hit a global. Plus there's the additional excitement below:

> I first started hitting this issue when using the new clientCertEngine parameter for node.js.

Hrm, yeah, that is awkward of a feature for Node to ship. You now even must worry if the OpenSSL Node statically links with is same as the OpenSSL that your engine is dynamically linked with. Node currently ship with 1.0.x, so an engine linked with 1.1.x will hit problems as they're not ABI-compatible. (Given your mention of `EC_KEY_METHOD_set_init`, I take it this is exactly your situation? Things are probably vaguely working only because the structs you use are probably still relatively close between 1.0.x and 1.1.x, but there's no guarantee they'd stay that way.)

Even two ABI-compatible versions will not work because of private structs. OpenSSL 1.1.0 opaquified `EVP_PKEY`, so it's valid for 1.1.0 => 1.1.1 or even 1.1.0 => 1.1.0a to change that struct. If that happens and you load both versions in a process, cross-contaminating an `EVP_PKEY` will interpret the pointer wrong and hit a memory error.

I should add that, even if you were to match versions with Node's 1.0.x release, 1.0.x has many more global locks than 1.1.x, so I expect even ref-counting won't be correctly locked.

Perhaps others OpenSSL contributors could chime in on how/whether `ENGINE` is intended to work in such a situation. @levitte? I am actually not familiar with `ENGINE`'s design, only with the pain and suffering that comes of mixing up two copies of a library in the same process. :-)

My sense is there's no way for this sort of static linking and dynamic engines work well. To add a feature of this shape to Node, one would need a less porous plugin interface than `ENGINE`, and then whoever consumes this plugin interface, Node or OpenSSL, would write the glue code to bridge things in.

**Comment by JackOfMostTrades** [comment]
Thanks for all of the insight @davidben! If anyone wants to chime in on best practices for using an ENGINE in this context that'd be helpful. In the mean time, node has no problem using a dynamically linked version of openssl if you build from source (and since it's also come up when I've been mucking about with python, you can do the same for the python's cryptography package).

I'll be leaving a comment on #1551 (since much of this was instigated by the parallel issue of providing a good-enough solution for hardware tokens that don't export the public component), but in the mean time I'd say you're right that this ticket probably isn't worth pursuing given that there seem to be a number of reasons why mixing static/dynamic linking isn't likely to work anyway.

**Comment by levitte** [comment]
An engine module should link with exactly the same libcrypto as the loading application.  Otherwise, as has already been mentioned, you will have issues when starting to pass around libcrypto objects.

The only way I can see when statically linking with libcrypto is to statically link with the engine modules as well.  When configuring OpenSSL with `no-shared`, the engines that come with OpenSSL will be included in libcrypto, and initialized properly as well.

The issue comes with many external engine modules, which have no mechanism whatsoever to be linked statically.  Some might, but that's nothing we've encouraged.
Of course, there might be some engine modules that find the libcrypto symbols from the calling program dynamically, but I'm unsure how many do this...

So in summary, trying to use dynamically loadable engine modules when statically linked with libcrypto is problematic and not recommended.

**Comment by JackOfMostTrades** [comment]
Cool, thanks for the insight everyone. I'm happy with a closed, won't fix resolution on this one.
